src/main/java/org/broadinstitute/hellbender/utils/downsampling/PositionalDownsampler.java

```diff
@@ -0,0 +1,146 @@
+package org.broadinstitute.hellbender.utils.downsampling;
+
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadConstants;
+import org.broadinstitute.hellbender.utils.read.ReadCoordinateComparator;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * PositionalDownsampler: Downsample each stack of reads at each alignment start to a size <= a target coverage
+ * using a {@link ReservoirDownsampler}. Stores only O(target coverage) reads in memory at any given time,
+ * provided the client regularly calls {@link #consumeFinalizedItems}.
+ *
+ * Unmapped reads with assigned positions are subject to downsampling in the same way as mapped reads,
+ * but unmapped reads without assigned positions are not subject to downsampling.
+ */
+public class PositionalDownsampler extends ReadsDownsampler {
+
+    private final ReservoirDownsampler reservoir;
+
+    private final SAMFileHeader header;
+
+    private GATKRead lastRead;
+
+    private List<GATKRead> finalizedReads;
+
+    /**
+     * Construct a PositionalDownsampler
+     *
+     * @param targetCoverage Maximum number of reads that may share any given alignment start position. Must be > 0
+     * @param header SAMFileHeader to use to determine contig ordering. Non-null.
+     */
+    public PositionalDownsampler( final int targetCoverage, final SAMFileHeader header ) {
+        Utils.validateArg(targetCoverage > 0, "targetCoverage must be > 0");
+        Utils.nonNull(header);
+
+        this.reservoir = new ReservoirDownsampler(targetCoverage);
+        this.finalizedReads = new ArrayList<>();
+        this.header = header;
+        clearItems();
+        resetStats();
+    }
+
+    @Override
+    public void submit( final GATKRead newRead ) {
+        // If we've moved to a new position, finalize the reads currently in our reservoir.
+        handlePositionalChange(newRead);
+
+        // Pass-through reads that have no assigned position, to avoid downsampling all unmapped reads
+        // to the targetCoverage. Unmapped reads that do have an assigned position *will* be subject to
+        // downsampling, however.
+        if ( readHasNoPosition(newRead) ) {
+            finalizedReads.add(newRead);
+        }
+        else {
+            final int reservoirPreviouslyDiscardedItems = reservoir.getNumberOfDiscardedItems();
+            reservoir.submit(newRead);
+            numDiscardedItems += reservoir.getNumberOfDiscardedItems() - reservoirPreviouslyDiscardedItems;
```

Comment:
Since that would change all the other downsamplers, I'd prefer to do it in a separate branch if you feel strongly about it. I think it's fine as-is, though.